vars/washeryCleanSnapshots.groovy

```diff
@@ -0,0 +1,111 @@
+/***********************************
+washeryCleanSnapshots DSL
+
+Cleans old washery snapshots from RDS Snapshot storage
+
+example usage
+washeryCleanSnapshots(
+  region: 'us-east-1', // (required, aws region to deploy the stack)
+  accountId: env.DEV_ACCOUNT_ID,
+  role: env.CIINABOXV2_ROLE, // IAM role to assume
+  type: 'rds|dbcluster',
+  snapshotRetainCount: 3 // The number of washery snapshots to keep stored
+)
+************************************/
+
+import com.base2.ciinabox.aws.AwsClientBuilder
+import com.amazonaws.services.rds.model.DescribeDBSnapshotsRequest
+import com.amazonaws.services.rds.model.DeleteDBSnapshotRequest
+import com.amazonaws.services.rds.model.DescribeDBClusterSnapshotsRequest
+import com.amazonaws.services.rds.model.DeleteDBClusterSnapshotRequest
+
+def call(body) {
+    def config = body
+    def clientBuilder = new AwsClientBuilder([
+        region: config.region,
+        awsAccountId: config.accountId,
+        role: config.role
+    ])  
+
+    def client = clientBuilder.rds()
+
+    if (config.type.toLowerCase() == 'rds') {
+        cleanInstanceWasherySnapshots(client, config.snapshotRetainCount)
+    } else if (config.type.toLowerCase() == 'dbcluster') {
+        cleanClusterWasherySnapshots(client, config.snapshotRetainCount)
+    } else {
+        throw new GroovyRuntimeException("washeryCleanSnapshots() doesn't support type ${config.type}")
+    }
+}
+
+@NonCPS
+def cleanInstanceWasherySnapshots(client, snapshotRetainCount){
+    //RDS Instance
+    def request = new DescribeDBSnapshotsRequest()
+            .withSnapshotType("manual")
+    def snapshotsResult = client.describeDBSnapshots(request)
+    def snapshots = snapshotsResult.getDBSnapshots()
+    def washerySnapshots = []
+
+    //Retrieve snapshots prefixed with `washery-scrubbed`
+    for (snapshot in snapshots) {
+        if (snapshot.getDBSnapshotIdentifier().startsWith("washery-scrubbed-")){
+            washerySnapshots.add(snapshot)
+        }
+    }
+
+    //Sort washery snapshots based on snapshot create time
+    washerySnapshots.sort{a,b-> b.getSnapshotCreateTime().compareTo(a.getSnapshotCreateTime())}
+    
+    //If retain count is less than the size of the total washery snapshots
+    if (snapshotRetainCount < washerySnapshots.size()){
+        def delete = washerySnapshots[snapshotRetainCount..-1]
+        println delete
+
+        //Delete snapshot's until only the snapshotRetainCount amount remains
```

Comment:
can you add a dry run option so we could run it first and see what is being deleted before we run the delete

Reply:
Thats a great idea.

Added dry run option (defaults to true).
```
 def dryRun =  config.get('dryRun',true)
```
```
//Delete snapshot's until only the snapshotRetainCount amount remains
        if (dryRun == false){
            for (snapshot in delete) {
                snapshot_identifier = snapshot.getDBSnapshotIdentifier()
                def delete_request = new DeleteDBSnapshotRequest().withDBSnapshotIdentifier(snapshot_identifier)
                def response = client.deleteDBSnapshot(delete_request)
                echo "Deleted Snapshot - ${snapshot_identifier} created on ${snapshot.getSnapshotCreateTime()}"
            }
        } 
```

Logs example:
```
Washery Snapshots to Keep - [{AvailabilityZones: [ap-southeast-2a, ap-southeast-2a, ap-southeast-2b, ap-southeast-2b, ap-southeast-2c, ap-southeast-2c],DBClusterSnapshotIdentifier: washery-scrubbed-aaaaaa,DBClusterIdentifier: washery-stack-test-rdscluster-eae0abz4wetd,SnapshotCreateTime: Tue Jan 24 02:38:05 UTC 2023,Engine: aurora,EngineMode: serverless,AllocatedStorage: 0,Status: available,Port: 0,VpcId: vpc-01887cba338bc68e7,ClusterCreateTime: Wed Jan 18 02:33:17 UTC 2023,MasterUsername: sldkfjosfsdfsdf,EngineVersion: 5.6.mysql_aurora.1.22.3,LicenseModel: aurora,SnapshotType: manual,PercentProgress: 100,StorageEncrypted: true,KmsKeyId: arn:aws:kms:ap-southeast-2:857301260320:key/faf7bec3-dc9a-4b9c-a08e-961fb67b9659,DBClusterSnapshotArn: arn:aws:rds:ap-southeast-2:857301260320:cluster-snapshot:washery-scrubbed-aaaaaa,IAMDatabaseAuthenticationEnabled: false,TagList: []}, {AvailabilityZones: [ap-southeast-2a, ap-southeast-2a, ap-southeast-2b, ap-southeast-2b, ap-southeast-2c, ap-southeast-2c],DBClusterSnapshotIdentifier: washery-scrubbed-lalksand,DBClusterIdentifier: washery-stack-test-rdscluster-eae0abz4wetd,SnapshotCreateTime: Tue Jan 24 02:30:27 UTC 2023,Engine: aurora,EngineMode: serverless,AllocatedStorage: 0,Status: available,Port: 0,VpcId: vpc-01887cba338bc68e7,ClusterCreateTime: Wed Jan 18 02:33:17 UTC 2023,MasterUsername: sldkfjosfsdfsdf,EngineVersion: 5.6.mysql_aurora.1.22.3,LicenseModel: aurora,SnapshotType: manual,PercentProgress: 100,StorageEncrypted: true,KmsKeyId: arn:aws:kms:ap-southeast-2:857301260320:key/faf7bec3-dc9a-4b9c-a08e-961fb67b9659,DBClusterSnapshotArn: arn:aws:rds:ap-southeast-2:857301260320:cluster-snapshot:washery-scrubbed-lalksand,IAMDatabaseAuthenticationEnabled: false,TagList: []}, {AvailabilityZones: [ap-southeast-2a, ap-southeast-2a, ap-southeast-2b, ap-southeast-2b, ap-southeast-2c, ap-southeast-2c],DBClusterSnapshotIdentifier: washery-scrubbed-testtestest,DBClusterIdentifier: washery-stack-test-rdscluster-eae0abz4wetd,SnapshotCreateTime: Mon Jan 23 23:59:19 UTC 2023,Engine: aurora,EngineMode: serverless,AllocatedStorage: 0,Status: available,Port: 0,VpcId: vpc-01887cba338bc68e7,ClusterCreateTime: Wed Jan 18 02:33:17 UTC 2023,MasterUsername: sldkfjosfsdfsdf,EngineVersion: 5.6.mysql_aurora.1.22.3,LicenseModel: aurora,SnapshotType: manual,PercentProgress: 100,StorageEncrypted: true,KmsKeyId: arn:aws:kms:ap-southeast-2:857301260320:key/faf7bec3-dc9a-4b9c-a08e-961fb67b9659,DBClusterSnapshotArn: arn:aws:rds:ap-southeast-2:857301260320:cluster-snapshot:washery-scrubbed-testtestest,IAMDatabaseAuthenticationEnabled: false,TagList: []}]
[Pipeline] echo
Washery snapshots to be Deleted - [{AvailabilityZones: [ap-southeast-2a, ap-southeast-2a, ap-southeast-2b, ap-southeast-2b, ap-southeast-2c, ap-southeast-2c],DBClusterSnapshotIdentifier: washery-scrubbed-jaja,DBClusterIdentifier: washery-stack-test-rdscluster-eae0abz4wetd,SnapshotCreateTime: Mon Jan 23 05:23:32 UTC 2023,Engine: aurora,EngineMode: serverless,AllocatedStorage: 0,Status: available,Port: 0,VpcId: vpc-01887cba338bc68e7,ClusterCreateTime: Wed Jan 18 02:33:17 UTC 2023,MasterUsername: sldkfjosfsdfsdf,EngineVersion: 5.6.mysql_aurora.1.22.3,LicenseModel: aurora,SnapshotType: manual,PercentProgress: 100,StorageEncrypted: true,KmsKeyId: arn:aws:kms:ap-southeast-2:857301260320:key/faf7bec3-dc9a-4b9c-a08e-961fb67b9659,DBClusterSnapshotArn: arn:aws:rds:ap-southeast-2:857301260320:cluster-snapshot:washery-scrubbed-jaja,IAMDatabaseAuthenticationEnabled: false,TagList: []}, {AvailabilityZones: [ap-southeast-2a, ap-southeast-2a, ap-southeast-2b, ap-southeast-2b, ap-southeast-2c, ap-southeast-2c],DBClusterSnapshotIdentifier: washery-scrubbed-testsnapkjsndkan,DBClusterIdentifier: washery-stack-test-rdscluster-eae0abz4wetd,SnapshotCreateTime: Mon Jan 23 05:21:14 UTC 2023,Engine: aurora,EngineMode: serverless,AllocatedStorage: 0,Status: available,Port: 0,VpcId: vpc-01887cba338bc68e7,ClusterCreateTime: Wed Jan 18 02:33:17 UTC 2023,MasterUsername: sldkfjosfsdfsdf,EngineVersion: 5.6.mysql_aurora.1.22.3,LicenseModel: aurora,SnapshotType: manual,PercentProgress: 100,StorageEncrypted: true,KmsKeyId: arn:aws:kms:ap-southeast-2:857301260320:key/faf7bec3-dc9a-4b9c-a08e-961fb67b9659,DBClusterSnapshotArn: arn:aws:rds:ap-southeast-2:857301260320:cluster-snapshot:washery-scrubbed-testsnapkjsndkan,IAMDatabaseAuthenticationEnabled: false,TagList: []}]
[Pipeline] echo
Dry run is set so no snapshots will be deleted
```